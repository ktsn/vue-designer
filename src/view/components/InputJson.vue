<template>
  <input class="input-json" type="text" :value="jsonValue" @change="onChange">
</template>

<script lang="ts">
import Vue from 'vue'

export default Vue.extend({
  name: 'InputJson',

  props: {
    value: [Number, String, Boolean, Object, Array]
  },

  computed: {
    jsonValue(): string {
      try {
        return this.value === undefined
          ? 'undefined'
          : JSON.stringify(this.value)
      } catch (e) {
        return 'undefined'
      }
    }
  },

  methods: {
    onChange(event: Event) {
      const input = event.target as HTMLInputElement
      const jsonValue = input.value

      try {
        const value =
          jsonValue === 'undefined' ? undefined : JSON.parse(jsonValue)

        this.$emit('change', value)
      } catch (e) {
        input.value = this.jsonValue
        this.$emit('error', jsonValue)
      }
    }
  }
})
</script>

<style lang="scss" scoped>
.input-json {
  padding: 0;
  border-width: 0;
  background: none;
  font-size: inherit;
  font-family: inherit;
}
</style>
