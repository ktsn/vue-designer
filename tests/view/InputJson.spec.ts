import { mount, Wrapper } from '@vue/test-utils'
import InputJson from '@/view/components/InputJson.vue'

describe('InputJson', () => {
  describe('Check render', () => {
    it('renders string value', () => {
      const wrapper = mount(InputJson, {
        propsData: {
          field: {
            name: 'stringName',
            value: 'string value'
          }
        }
      })

      expect(wrapper.html()).toMatchSnapshot()
    })

    it('renders number value', () => {
      const wrapper = mount(InputJson, {
        propsData: {
          field: {
            name: 'numberName',
            value: 123
          }
        }
      })

      expect(wrapper.html()).toMatchSnapshot()
    })

    it('renders boolean value', () => {
      const wrapper = mount(InputJson, {
        propsData: {
          field: {
            name: 'booleanName',
            value: true
          }
        }
      })

      expect(wrapper.html()).toMatchSnapshot()
    })

    it('renders null value', () => {
      const wrapper = mount(InputJson, {
        propsData: {
          field: {
            name: 'nullName',
            value: null
          }
        }
      })

      expect(wrapper.html()).toMatchSnapshot()
    })

    it('renders undefined value', () => {
      const wrapper = mount(InputJson, {
        propsData: {
          field: {
            name: 'numberName',
            value: undefined
          }
        }
      })

      expect(wrapper.html()).toMatchSnapshot()
    })

    it('renders object value', () => {
      const wrapper = mount(InputJson, {
        propsData: {
          field: {
            name: 'objectName',
            value: {
              foo: 'test',
              bar: 123
            }
          }
        }
      })

      expect(wrapper.html()).toMatchSnapshot()
    })

    it('renders array value', () => {
      const wrapper = mount(InputJson, {
        propsData: {
          field: {
            name: 'arrayName',
            value: ['test', 123, true]
          }
        }
      })

      expect(wrapper.html()).toMatchSnapshot()
    })
  })

  describe('Behavior', () => {
    it('notifies updated value', () => {
      const wrapper = mount(InputJson, {
        propsData: {
          field: {
            name: 'test',
            value: 123
          }
        }
      })
      const ctrl = new InputJsonController(wrapper)

      ctrl.clickEdit()
      ctrl.editValue('456')
      ctrl.clickApply()

      expect(wrapper.emitted('change')[0]).toEqual([
        {
          name: 'test',
          value: 456
        }
      ])
    })

    it('converts json value', () => {
      const wrapper = mount(InputJson, {
        propsData: {
          field: {
            name: 'test',
            value: 123
          }
        }
      })
      const ctrl = new InputJsonController(wrapper)

      ctrl.clickEdit()
      ctrl.editValue('{"foo": "123", "bar": 123, "baz": null}')
      ctrl.clickApply()

      expect(wrapper.emitted('change')[0]).toEqual([
        {
          name: 'test',
          value: {
            foo: '123',
            bar: 123,
            baz: null
          }
        }
      ])
    })

    it('edits child property value', () => {
      const wrapper = mount(InputJson, {
        propsData: {
          field: {
            name: 'test',
            value: {
              foo: 'abc',
              bar: 123
            }
          }
        }
      })
      const bar = wrapper.findAll(InputJson).at(2)
      const ctrl = new InputJsonController(bar)

      ctrl.clickEdit()
      ctrl.editValue('456')
      ctrl.clickApply()

      expect(wrapper.emitted('change')[0]).toEqual([
        {
          name: 'test',
          value: {
            foo: 'abc',
            bar: 456
          }
        }
      ])
    })

    it('edits array item', () => {
      const wrapper = mount(InputJson, {
        propsData: {
          field: {
            name: 'test',
            value: ['foo', 123, true]
          }
        }
      })
      const second = wrapper.findAll(InputJson).at(2)
      const ctrl = new InputJsonController(second)

      ctrl.clickEdit()
      ctrl.editValue('456')
      ctrl.clickApply()

      expect(wrapper.emitted('change')[0]).toEqual([
        {
          name: 'test',
          value: ['foo', 456, true]
        }
      ])
    })

    it('rejects invalid json string', () => {
      const wrapper = mount<any>(InputJson, {
        propsData: {
          field: {
            name: 'test',
            value: 123
          }
        }
      })
      const ctrl = new InputJsonController(wrapper)

      ctrl.clickEdit()
      ctrl.editValue('abc')
      ctrl.clickApply()

      expect(wrapper.emitted('change')).toBe(undefined)
    })

    it('is editable child property name', () => {
      const wrapper = mount(InputJson, {
        propsData: {
          field: {
            name: 'test',
            value: {
              foo: 'abc',
              bar: 'def'
            }
          }
        }
      })
      const child = wrapper.findAll(InputJson).at(1)
      const ctrl = new InputJsonController(child)

      ctrl.clickEdit()
      ctrl.editName('baz')
      ctrl.clickApply()

      expect(wrapper.emitted('change')[0]).toEqual([
        {
          name: 'test',
          value: {
            bar: 'def',
            baz: 'abc'
          }
        }
      ])
    })

    it('is not editable root property name', () => {
      const wrapper = mount(InputJson, {
        propsData: {
          field: {
            name: 'test',
            value: 123
          }
        }
      })
      const ctrl = new InputJsonController(wrapper)

      ctrl.clickEdit()
      ctrl.editName('test2')
      ctrl.clickApply()

      expect(wrapper.emitted('change')[0]).toEqual([
        {
          name: 'test',
          value: 123
        }
      ])
    })

    it('is not editable array property name', () => {
      const wrapper = mount(InputJson, {
        propsData: {
          field: {
            name: 'test',
            value: ['foo', 'bar', 'baz']
          }
        }
      })
      const child = wrapper.find(InputJson)
      const ctrl = new InputJsonController(child)

      ctrl.clickEdit()
      ctrl.editName('3')
      ctrl.clickApply()

      expect(wrapper.emitted('change')[0]).toEqual([
        {
          name: 'test',
          value: ['foo', 'bar', 'baz']
        }
      ])
    })
  })
})

class InputJsonController {
  constructor(private wrapper: Wrapper<any>) {}

  clickEdit(): void {
    if (!this.wrapper.vm.editing) {
      this.wrapper.find('.input-json-button').element.click()
    }
  }

  clickApply(): void {
    if (this.wrapper.vm.editing) {
      this.wrapper.find('.input-json-button:nth-child(2)').element.click()
    }
  }

  editName(str: string): void {
    const input = this.wrapper.find('.input-json-label.editing')
      .element as HTMLInputElement
    if (input) {
      input.value = str
      input.dispatchEvent(new Event('input'))
    }
  }

  editValue(str: string): void {
    const input = this.wrapper.find('.input-json-value.editing')
      .element as HTMLInputElement
    if (input) {
      input.value = str
      input.dispatchEvent(new Event('input'))
    }
  }
}
