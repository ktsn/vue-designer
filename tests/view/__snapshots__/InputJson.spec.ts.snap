// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`InputJson > Check render > renders array value 1`] = `
<div class="input-json" data-test-id="input-json">
  <div class="input-json-line"><!-- Object property key --><span class="input-json-label">arrayName</span><!-- Object value --><span class="array input-json-value">Array[3]</span><!-- Actions -->
    <div class="input-json-actions"><button type="button" class="input-json-button edit" aria-label="Edit"><svg class="base-icon input-json-icon">
          <use href="#ic_create_24px"></use>
        </svg></button></div>
  </div><!-- Children -->
  <ul class="input-json-children">
    <li>
      <div class="input-json" data-test-id="input-json">
        <div class="input-json-line"><!-- Object property key --><span class="input-json-label">0</span><!-- Object value --><span class="string input-json-value">"test"</span><!-- Actions -->
          <div class="input-json-actions"><button type="button" class="input-json-button edit" aria-label="Edit"><svg class="base-icon input-json-icon">
                <use href="#ic_create_24px"></use>
              </svg></button></div>
        </div><!-- Children --><!--v-if-->
      </div>
    </li>
    <li>
      <div class="input-json" data-test-id="input-json">
        <div class="input-json-line"><!-- Object property key --><span class="input-json-label">1</span><!-- Object value --><span class="number input-json-value">123</span><!-- Actions -->
          <div class="input-json-actions"><button type="button" class="input-json-button edit" aria-label="Edit"><svg class="base-icon input-json-icon">
                <use href="#ic_create_24px"></use>
              </svg></button></div>
        </div><!-- Children --><!--v-if-->
      </div>
    </li>
    <li>
      <div class="input-json" data-test-id="input-json">
        <div class="input-json-line"><!-- Object property key --><span class="input-json-label">2</span><!-- Object value --><span class="boolean input-json-value">true</span><!-- Actions -->
          <div class="input-json-actions"><button type="button" class="input-json-button edit" aria-label="Edit"><svg class="base-icon input-json-icon">
                <use href="#ic_create_24px"></use>
              </svg></button></div>
        </div><!-- Children --><!--v-if-->
      </div>
    </li>
  </ul>
</div>
`;

exports[`InputJson > Check render > renders boolean value 1`] = `
<div class="input-json" data-test-id="input-json">
  <div class="input-json-line"><!-- Object property key --><span class="input-json-label">booleanName</span><!-- Object value --><span class="boolean input-json-value">true</span><!-- Actions -->
    <div class="input-json-actions"><button type="button" class="input-json-button edit" aria-label="Edit"><svg class="base-icon input-json-icon">
          <use href="#ic_create_24px"></use>
        </svg></button></div>
  </div><!-- Children --><!--v-if-->
</div>
`;

exports[`InputJson > Check render > renders null value 1`] = `
<div class="input-json" data-test-id="input-json">
  <div class="input-json-line"><!-- Object property key --><span class="input-json-label">nullName</span><!-- Object value --><span class="null input-json-value">null</span><!-- Actions -->
    <div class="input-json-actions"><button type="button" class="input-json-button edit" aria-label="Edit"><svg class="base-icon input-json-icon">
          <use href="#ic_create_24px"></use>
        </svg></button></div>
  </div><!-- Children --><!--v-if-->
</div>
`;

exports[`InputJson > Check render > renders number value 1`] = `
<div class="input-json" data-test-id="input-json">
  <div class="input-json-line"><!-- Object property key --><span class="input-json-label">numberName</span><!-- Object value --><span class="number input-json-value">123</span><!-- Actions -->
    <div class="input-json-actions"><button type="button" class="input-json-button edit" aria-label="Edit"><svg class="base-icon input-json-icon">
          <use href="#ic_create_24px"></use>
        </svg></button></div>
  </div><!-- Children --><!--v-if-->
</div>
`;

exports[`InputJson > Check render > renders object value 1`] = `
<div class="input-json" data-test-id="input-json">
  <div class="input-json-line"><!-- Object property key --><span class="input-json-label">objectName</span><!-- Object value --><span class="object input-json-value">Object</span><!-- Actions -->
    <div class="input-json-actions"><button type="button" class="input-json-button edit" aria-label="Edit"><svg class="base-icon input-json-icon">
          <use href="#ic_create_24px"></use>
        </svg></button></div>
  </div><!-- Children -->
  <ul class="input-json-children">
    <li>
      <div class="input-json" data-test-id="input-json">
        <div class="input-json-line"><!-- Object property key --><span class="input-json-label">foo</span><!-- Object value --><span class="string input-json-value">"test"</span><!-- Actions -->
          <div class="input-json-actions"><button type="button" class="input-json-button edit" aria-label="Edit"><svg class="base-icon input-json-icon">
                <use href="#ic_create_24px"></use>
              </svg></button></div>
        </div><!-- Children --><!--v-if-->
      </div>
    </li>
    <li>
      <div class="input-json" data-test-id="input-json">
        <div class="input-json-line"><!-- Object property key --><span class="input-json-label">bar</span><!-- Object value --><span class="number input-json-value">123</span><!-- Actions -->
          <div class="input-json-actions"><button type="button" class="input-json-button edit" aria-label="Edit"><svg class="base-icon input-json-icon">
                <use href="#ic_create_24px"></use>
              </svg></button></div>
        </div><!-- Children --><!--v-if-->
      </div>
    </li>
  </ul>
</div>
`;

exports[`InputJson > Check render > renders string value 1`] = `
<div class="input-json" data-test-id="input-json">
  <div class="input-json-line"><!-- Object property key --><span class="input-json-label">stringName</span><!-- Object value --><span class="string input-json-value">"string value"</span><!-- Actions -->
    <div class="input-json-actions"><button type="button" class="input-json-button edit" aria-label="Edit"><svg class="base-icon input-json-icon">
          <use href="#ic_create_24px"></use>
        </svg></button></div>
  </div><!-- Children --><!--v-if-->
</div>
`;

exports[`InputJson > Check render > renders undefined value 1`] = `
<div class="input-json" data-test-id="input-json">
  <div class="input-json-line"><!-- Object property key --><span class="input-json-label">numberName</span><!-- Object value --><span class="undefined input-json-value">undefined</span><!-- Actions -->
    <div class="input-json-actions"><button type="button" class="input-json-button edit" aria-label="Edit"><svg class="base-icon input-json-icon">
          <use href="#ic_create_24px"></use>
        </svg></button></div>
  </div><!-- Children --><!--v-if-->
</div>
`;
